File: Strand/Data/CycleTrackingStore.swift
import Foundation
import WhoopStore

// MARK: - Menstrual cycle tracker storage
//
// A logged period start is one value-1 metricSeries point under the dedicated local source
// `noop-cycle`. The source is intentionally separate from strap/import/computed ids so an import or
// re-analysis cannot overwrite sensitive user-entered history. This is the byte-parity twin of Android
// `CycleTrackingStore`: same source id, key, value, local ISO day, ordering, and real-delete semantics.
// No schema change is needed and no data leaves the device.

enum CycleTrackingStore {
    static let sourceId = "noop-cycle"
    static let periodStartKey = "period_start"
    static let loggedValue = 1.0
    static let earliestDay = "0000-01-01"
    static let latestDay = "9999-12-31"
}

extension Repository {

    /// Logged period-start days, oldest first. A value gate makes reads forward-compatible if another
    /// cycle event kind is ever projected into this source.
    func periodStarts(from: String = CycleTrackingStore.earliestDay,
                      to: String = CycleTrackingStore.latestDay) async -> [String] {
        guard let store = await storeHandle() else { return [] }
        let rows = (try? await store.metricSeries(deviceId: CycleTrackingStore.sourceId,
                                                  key: CycleTrackingStore.periodStartKey,
                                                  from: from, to: to)) ?? []
        return rows.filter { $0.value >= CycleTrackingStore.loggedValue }.map(\.day)
    }

    /// Log (or idempotently re-log) one local calendar day as cycle day 1.
    func logPeriodStart(day: String) async {
        guard let store = await storeHandle() else { return }
        _ = try? await store.upsertMetricSeries(
            [MetricPoint(day: day, key: CycleTrackingStore.periodStartKey,
                         value: CycleTrackingStore.loggedValue)],
            deviceId: CycleTrackingStore.sourceId)
        noteCycleTrackingChanged()
    }

    /// Remove one logged start. This is a physical row delete, not a zero/sentinel marker.
    func deletePeriodStart(day: String) async {
        guard let store = await storeHandle() else { return }
        _ = try? await store.deleteMetricSeriesPoint(deviceId: CycleTrackingStore.sourceId,
                                                      day: day,
                                                      key: CycleTrackingStore.periodStartKey)
        noteCycleTrackingChanged()
    }

    /// Delete every logged period start after an explicit user confirmation.
    func deleteAllPeriodStarts() async {
        guard let store = await storeHandle() else { return }
        _ = try? await store.deleteMetricSeries(deviceId: CycleTrackingStore.sourceId,
                                                 key: CycleTrackingStore.periodStartKey)
        noteCycleTrackingChanged()
    }
}

File: android/app/src/main/java/com/noop/data/CycleTrackingStore.kt
package com.noop.data

import java.time.LocalDate

/**
 * Local-only menstrual cycle tracker storage. A period start is one value-1 [MetricSeriesRow] under
 * the dedicated `noop-cycle` source. This is the value-for-value twin of Swift
 * `CycleTrackingStore`: same source, key, day format, ordering, idempotent upsert and physical delete.
 * Imports and strap analysis use different source ids, so they cannot overwrite user-entered history.
 */
class CycleTrackingStore(
    private val upsertRows: suspend (List<MetricSeriesRow>) -> Unit,
    private val queryRows: suspend (String, String, String, String) -> List<MetricSeriesRow>,
    private val deletePoint: suspend (String, String, String) -> Unit,
    private val deleteSeries: suspend (String, String) -> Unit,
) {
    constructor(repo: WhoopRepository) : this(
        { rows -> repo.upsertMetricSeries(rows) },
        { deviceId, key, from, to -> repo.metricSeries(deviceId, key, from, to) },
        { deviceId, day, key -> repo.deleteMetricSeriesPoint(deviceId, day, key) },
        { deviceId, key -> repo.deleteMetricSeries(deviceId, key) },
    )

    /** Log or idempotently re-log cycle day 1. */
    suspend fun logStart(day: String = todayKey()) {
        upsertRows(listOf(MetricSeriesRow(SOURCE_ID, day, PERIOD_START_KEY, LOGGED_VALUE)))
    }

    /** Logged starts in the inclusive range, oldest first. */
    suspend fun starts(from: String = DAY_MIN, to: String = DAY_MAX): List<String> =
        queryRows(SOURCE_ID, PERIOD_START_KEY, from, to)
            .filter { it.value >= LOGGED_VALUE }
            .sortedBy { it.day }
            .map { it.day }

    /** Physically remove one start row; deleting an absent row is an idempotent no-op. */
    suspend fun deleteStart(day: String) {
        deletePoint(SOURCE_ID, day, PERIOD_START_KEY)
    }

    /** Remove all logged starts after explicit confirmation in the UI. */
    suspend fun deleteAll() {
        deleteSeries(SOURCE_ID, PERIOD_START_KEY)
    }

    companion object {
        const val SOURCE_ID = "noop-cycle"
        const val PERIOD_START_KEY = "period_start"
        const val LOGGED_VALUE = 1.0
        private const val DAY_MIN = "0000-01-01"
        private const val DAY_MAX = "9999-12-31"

        fun todayKey(today: LocalDate = LocalDate.now()): String = today.toString()
    }
}
